components/ulp/esp32ulp_mapgen.py

```diff
@@ -2,71 +2,95 @@
-    f_ld.write(textwrap.dedent(
-        f"""
-        /* ULP variable definitions for the linker.
-         * This file is generated automatically by {UTIL} utility.
-         */
-        """
-    ))
-    f_h.write(textwrap.dedent(
-        f"""
-        /* ULP variable definitions for the compiler.
-         * This file is generated automatically by {UTIL} utility.
-         */
-        #pragma once
-        #ifdef __cplusplus
-        extern "C" {{
-        #endif
-        """
-    ))
+def name_mangling(name: str) -> str:
+    # Simple and dumb name mangling for namespaced name following GCC algorithm
+    ns, n = name.split('::')
+    return '_ZN{0}{1}{2}{3}E'.format(len(ns), ns, len(n), n)
+
+
+def gen_h_from_sym(f_sym: typing.TextIO, f_h: typing.TextIO, base_addr: int, prefix: str) -> None:
+    cpp_mode = False
+    if '::' in prefix:
+        # C++ mode, let's avoid the extern "C" type and instead use namespace
+        f_h.write(textwrap.dedent(
+            f"""
+            /* ULP variable definitions for the compiler.
+             * This file is generated automatically by {UTIL} utility.
+             */
+            #pragma once
+            """  # noqa: E222
+        ))
+
+        f_h.write('namespace {0} {{\n'.format(prefix.split('::')[0]))
+        cpp_mode = True
+    else:
+        f_h.write(textwrap.dedent(
+            f"""
+            /* ULP variable definitions for the compiler.
+             * This file is generated automatically by {UTIL} utility.
+             */
+            #pragma once
+            #ifdef __cplusplus
+            extern "C" {{
+            #endif
+            """  # noqa: E222
+        ))
 
+    expr = re.compile('^\\s*\\d+: ([a-f0-9]{8})\\s+(\\d+) OBJECT\\s+GLOBAL\\s+DEFAULT\\s+[^ ]+ (.*)$')
     for line in f_sym:
-        # NM "posix" format output has the following structure:
-        # symbol_name symbol_type addr_hex [size_hex]
-        parts = line.split()
-        name = parts[0]
-        addr = int(parts[2], 16) + base_addr
-        f_h.write('extern uint32_t ulp_{0};\n'.format(name))
-        f_ld.write('PROVIDE ( ulp_{0} = 0x{1:08x} );\n'.format(name, addr))
+        # readelf format output has the following structure:
+        #   index: addr_hex     size TYPE  SCOPE DEFAULT    junk symbol_name
+        # So match the line with a regular expression to parse it first
+        groups = expr.match(line)
+        if groups is None:  # Ignore non global or non object
+            continue
+        addr = int(groups.group(1), 16) + base_addr
+        size = int(groups.group(2))
+        name = groups.group(3)
+        if not cpp_mode:
+            name = prefix + name
+        f_h.write('extern uint32_t {0}{1};'.format(name, '[{0}]'.format(int(size / 4)) if size > 4 else ''))
+        # Then set the address to find the variable
+        f_h.write(' asm(".equ {0}, 0x{1:08X}");\n'.format(name_mangling(prefix + name) if cpp_mode else name, addr))
```

Comment:
Compared to generation of the linker script, seems like this approach might prevent generation of bindings for programs written in other languages, e.g. Rust.

Reply:
> The asm block will only have effect if it is included into a C or C++ source file and processed by a C or C++ compiler. If a bindings generator parses the header file, instead, then the asm block won't have any effect.

That's true but I don't see how it would be different from the linker's script. If the binding generators parses the C header string as it is now, it doesn't know the address for the symbol either. The only way for the binding generator to know the address for a symbol right now is to parse the linker's script, which is a PITA to get right. It's probably a lot easier to parse the output of `nm` or `readelf`

The issue with the linker's script is that it doesn't choke on error. The linker will silently accept 2 variables with the same name and different addresses and use whatever comes first or last in the linking order, IDK. I've lost a complete day figuring out why the ULP didn't worked the way I expected thinking that the ULP code or HP code was wrong, but the error was in the tool not the code. 

When the address is set in the C header file, if you have 2 similar declarations with different address in the same translation unit (that will happen in your main.cpp file for example), then you'll get a compile time error, IMHO it's much better.

Reply:
Nevermind, I've just tested it and it didn't choke as I thought it would. The asm directive just defines a variable/constant and the code that's using the variable get replaced by the constant. It accepts that the constant is changed, unlike what's written in the documentation. So there's no reason to use the asm directive in the code, I'll reinstantiate the linker's script then. I need to find a way to get a linker error if one symbol is declared twice. I think it can be done by dropping the "PROVIDE" part of the line, so that multiple definition will emit an error. I need to test to make sure it works as intended.

Reply:
Dropped that part and finally used a linker script with some additional protection to prevent linking in case the symbol is already defined.